File: LeetCode/cpp/496_next_greater_element_i.cc
/*
Difficulty: Easy
*/

#include <cstddef>
#include <unordered_map>
#include <utility>
#include <vector>

class Solution {
 public:
  /// Time: O(n + m), Space: O(n)
  /// n = nums1.size(), m = nums2.size(), n <= m
  std::vector<int> nextGreaterElement(std::vector<int>& nums1,
                                      const std::vector<int>& nums2) {
    // value -> next greater value
    std::unordered_map<int, int> nextGreater;
    for (const int num : nums1) {
      nextGreater[num] = -1;
    }

    std::vector<int> stack;
    for (const int currNum : nums2) {
      while (!stack.empty() && stack.back() < currNum) {
        nextGreater[stack.back()] = currNum;
        stack.pop_back();
      }
      stack.push_back(currNum);
    }

    // Replace values in nums1 with next greatest value from nums2.
    for (std::size_t i = 0; i < nums1.size(); ++i) {
      nums1[i] = nextGreater.at(nums1[i]);
    }

    // Signature does not describe nums1 as const&, so can modify + move it.
    // In normal code, would instead make const& and return a new vector.
    return std::move(nums1);
  }
};

File: LeetCode/rust/496_next_greater_element_i.rs
/*
Difficulty: Easy
*/

use std::collections::HashMap;

impl Solution {
    /// Time: O(n + m), Space: (m)
    /// n = nums1.len, m = nums2.len
    pub fn next_greater_element(mut nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        // value -> next greater value coming from a greater index.
        let mut next_greater = HashMap::<i32, i32>::new();
        let mut stack = Vec::<i32>::new();

        // Find the next greater number for each number in nums2.
        for curr_num in nums2 {
            while let Some(&most_recent_num) = stack.last() {
                if most_recent_num < curr_num {
                    next_greater.insert(most_recent_num, curr_num);
                    stack.pop();
                } else {
                    break;
                }
            }
            stack.push(curr_num);
        }

        // Replace each value in nums1 with its next greater value in nums2.
        let next_greater = next_greater;
        for num in &mut nums1 {
            *num = next_greater.get(&num).copied().unwrap_or(-1);
        }
        nums1
    }
}
